File: src/main/java/com/dongtronic/diabot/converters/A1cConverter.java
package com.dongtronic.diabot.converters;

import com.dongtronic.diabot.data.A1cDTO;
import com.dongtronic.diabot.data.ConversionDTO;
import com.dongtronic.diabot.exceptions.UnknownUnitException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A1c conversion logic
 */
public class A1cConverter {

  private static final Logger logger = LoggerFactory.getLogger(A1cConverter.class);

  public static A1cDTO estimateA1c(String originalValue, String unit) throws UnknownUnitException {
    logger.info("Estimating A1c for BG " + originalValue);

    ConversionDTO glucoseConversionResult = BloodGlucoseConverter.convert(originalValue, unit);

    return estimateA1c(glucoseConversionResult);
  }

  public static A1cDTO estimateAverage(String originalValue) {
    double a1c = Double.valueOf(originalValue);

    if (a1c < 0 || a1c > 375) {
      throw new IllegalArgumentException();
    }

    A1cDTO result = null;

    try {
      if (a1c < 25) {
        result = estimateAverageDcct(a1c);
      } else {
        result = estimateAverageIfcc(a1c);
      }
    } catch (UnknownUnitException e) {
      // Ignored on purpose
    }

    return result;
  }

  private static A1cDTO estimateAverageDcct(double dcct) throws UnknownUnitException {
    double mgdl = convertDcctToMgdl(dcct);
    double ifcc = convertDcctToIfcc(dcct);

    ConversionDTO conversion = BloodGlucoseConverter.convert(String.valueOf(mgdl), "mgdl");

    return new A1cDTO(conversion, dcct, ifcc, 0, 0);
  }

  private static A1cDTO estimateAverageIfcc(double ifcc) throws UnknownUnitException {
    double mgdl = convertIfccToMgdl(ifcc);
    double dcct = convertIfccToDcct(ifcc);

    ConversionDTO conversion = BloodGlucoseConverter.convert(String.valueOf(mgdl), "mgdl");

    return new A1cDTO(conversion, dcct, ifcc, 0, 0);
  }


  private static A1cDTO estimateA1c(ConversionDTO glucose) {
    double ifcc_mgdl = 0;
    double dcct_mgdl = 0;
    double ifcc_mmol = 0;
    double dcct_mmol = 0;

    if (glucose.getInputUnit() == GlucoseUnit.MGDL) {
      ifcc_mgdl = convertMgdlToIfcc(glucose.getOriginal());
      dcct_mgdl = convertMgdlToDcct(glucose.getOriginal());
    } else if (glucose.getInputUnit() == GlucoseUnit.MMOL) {
      ifcc_mmol = convertMgdlToIfcc(glucose.getConverted());
      dcct_mmol = convertMgdlToDcct(glucose.getConverted());
    } else {
      return estimateA1cAmbiguous(glucose);
    }

    return new A1cDTO(glucose, dcct_mgdl, ifcc_mgdl, dcct_mmol, ifcc_mmol);
  }

  private static A1cDTO estimateA1cAmbiguous(ConversionDTO glucose) {

    double ifcc_mgdl = convertMgdlToIfcc(glucose.getMgdl());
    double dcct_mgdl = convertMgdlToDcct(glucose.getMgdl());
    double ifcc_mmol = convertMmolToIfcc(glucose.getMmol());
    double dcct_mmol = convertMmolToDcct(glucose.getMmol());

    return new A1cDTO(glucose, dcct_mgdl, ifcc_mgdl, dcct_mmol, ifcc_mmol);
  }

  private static double convertMmolToDcct(double glucose) {
    return convertMgdlToDcct(glucose * 18.016);
  }

  private static double convertMmolToIfcc(double glucose) {
    return convertMgdlToIfcc(glucose * 18.016);
  }

  private static double convertMgdlToDcct(double glucose) {
    return (glucose + 46.7) / 28.7;
  }

  private static double convertMgdlToIfcc(double glucose) {
    return (convertMgdlToDcct(glucose) - 2.15) * 10.929;
  }

  private static double convertDcctToMgdl(double dcct) {
    return (dcct * 28.7) - 46.7;
  }

  private static double convertIfccToMgdl(double ifcc) {
    return convertDcctToMgdl(convertIfccToDcct(ifcc));
  }

  private static double convertIfccToDcct(double ifcc) {
    return (ifcc / 10.929) + 2.15;
  }

  private static double convertDcctToIfcc(double dcct) {
    return (dcct - 2.15) * 10.929;
  }
}

File: src/main/java/com/dongtronic/diabot/converters/A1cConverter.kt
package com.dongtronic.diabot.converters

import com.dongtronic.diabot.data.A1cDTO
import com.dongtronic.diabot.data.ConversionDTO
import com.dongtronic.diabot.exceptions.UnknownUnitException
import org.slf4j.LoggerFactory

/**
 * A1c conversion logic
 */
object A1cConverter {

    private val logger = LoggerFactory.getLogger(A1cConverter::class.java)

    @Throws(UnknownUnitException::class)
    fun estimateA1c(originalValue: String, unit: String?): A1cDTO {
        logger.info("Estimating A1c for BG $originalValue")

        val glucoseConversionResult = BloodGlucoseConverter.convert(originalValue, unit)

        return estimateA1c(glucoseConversionResult!!)
    }

    fun estimateAverage(originalValue: String): A1cDTO? {
        val a1c = java.lang.Double.valueOf(originalValue)

        if (a1c < 0 || a1c > 375) {
            throw IllegalArgumentException()
        }

        var result: A1cDTO? = null

        try {
            result = if (a1c < 25) {
                estimateAverageDcct(a1c)
            } else {
                estimateAverageIfcc(a1c)
            }
        } catch (e: UnknownUnitException) {
            // Ignored on purpose
        }

        return result
    }

    @Throws(UnknownUnitException::class)
    private fun estimateAverageDcct(dcct: Double): A1cDTO {
        val mgdl = convertDcctToMgdl(dcct)
        val ifcc = convertDcctToIfcc(dcct)

        val conversion = BloodGlucoseConverter.convert(mgdl.toString(), "mgdl")

        return A1cDTO(conversion!!, dcct, ifcc, 0.0, 0.0)
    }

    @Throws(UnknownUnitException::class)
    private fun estimateAverageIfcc(ifcc: Double): A1cDTO {
        val mgdl = convertIfccToMgdl(ifcc)
        val dcct = convertIfccToDcct(ifcc)

        val conversion = BloodGlucoseConverter.convert(mgdl.toString(), "mgdl")

        return A1cDTO(conversion!!, dcct, ifcc, 0.0, 0.0)
    }


    private fun estimateA1c(glucose: ConversionDTO): A1cDTO {
        var ifccMgdl = 0.0
        var dcctMgdl = 0.0
        var ifccMmol = 0.0
        var dcctMmol = 0.0

        when {
            glucose.inputUnit == GlucoseUnit.MGDL -> {
                ifccMgdl = convertMgdlToIfcc(glucose.original)
                dcctMgdl = convertMgdlToDcct(glucose.original)
            }
            glucose.inputUnit == GlucoseUnit.MMOL -> {
                ifccMmol = convertMgdlToIfcc(glucose.converted)
                dcctMmol = convertMgdlToDcct(glucose.converted)
            }
            else -> return estimateA1cAmbiguous(glucose)
        }

        return A1cDTO(glucose, dcctMgdl, ifccMgdl, dcctMmol, ifccMmol)
    }

    private fun estimateA1cAmbiguous(glucose: ConversionDTO): A1cDTO {

        val ifccMgdl = convertMgdlToIfcc(glucose.mgdl.toDouble())
        val dcctMgdl = convertMgdlToDcct(glucose.mgdl.toDouble())
        val ifccMmol = convertMmolToIfcc(glucose.mmol)
        val dcctMmol = convertMmolToDcct(glucose.mmol)

        return A1cDTO(glucose, dcctMgdl, ifccMgdl, dcctMmol, ifccMmol)
    }

    private fun convertMmolToDcct(glucose: Double): Double {
        return convertMgdlToDcct(glucose * 18.016)
    }

    private fun convertMmolToIfcc(glucose: Double): Double {
        return convertMgdlToIfcc(glucose * 18.016)
    }

    private fun convertMgdlToDcct(glucose: Double): Double {
        return (glucose + 46.7) / 28.7
    }

    private fun convertMgdlToIfcc(glucose: Double): Double {
        return (convertMgdlToDcct(glucose) - 2.15) * 10.929
    }

    private fun convertDcctToMgdl(dcct: Double): Double {
        return dcct * 28.7 - 46.7
    }

    private fun convertIfccToMgdl(ifcc: Double): Double {
        return convertDcctToMgdl(convertIfccToDcct(ifcc))
    }

    private fun convertIfccToDcct(ifcc: Double): Double {
        return ifcc / 10.929 + 2.15
    }

    private fun convertDcctToIfcc(dcct: Double): Double {
        return (dcct - 2.15) * 10.929
    }
}
